crates/compiler/gen_dev/src/generic64/x86_64.rs

```diff
@@ -1452,6 +1452,83 @@ impl Assembler<X86_64GeneralReg, X86_64FloatReg> for X86_64Assembler {
     fn xor_reg64_reg64_reg64(buf: &mut Vec<'_, u8>, dst: Reg64, src1: Reg64, src2: Reg64) {
         binop_move_src_to_dst_reg64(buf, xor_reg64_reg64, dst, src1, src2)
     }
+
+    fn shl_reg64_reg64_reg64<'a, 'r, ASM, CC>(
+        buf: &mut Vec<'a, u8>,
+        storage_manager: &mut StorageManager<'a, 'r, X86_64GeneralReg, X86_64FloatReg, ASM, CC>,
+        dst: X86_64GeneralReg,
+        src1: X86_64GeneralReg,
+        src2: X86_64GeneralReg,
+    ) where
+        ASM: Assembler<X86_64GeneralReg, X86_64FloatReg>,
+        CC: CallConv<X86_64GeneralReg, X86_64FloatReg, ASM>,
+    {
+        shift_reg64_reg64_reg64(buf, storage_manager, shl_reg64_reg64, dst, src1, src2)
+    }
+
+    fn shr_reg64_reg64_reg64<'a, 'r, ASM, CC>(
+        buf: &mut Vec<'a, u8>,
+        storage_manager: &mut StorageManager<'a, 'r, X86_64GeneralReg, X86_64FloatReg, ASM, CC>,
+        dst: X86_64GeneralReg,
+        src1: X86_64GeneralReg,
+        src2: X86_64GeneralReg,
+    ) where
+        ASM: Assembler<X86_64GeneralReg, X86_64FloatReg>,
+        CC: CallConv<X86_64GeneralReg, X86_64FloatReg, ASM>,
+    {
+        shift_reg64_reg64_reg64(buf, storage_manager, shr_reg64_reg64, dst, src1, src2)
+    }
+
+    fn sar_reg64_reg64_reg64<'a, 'r, ASM, CC>(
+        buf: &mut Vec<'a, u8>,
+        storage_manager: &mut StorageManager<'a, 'r, X86_64GeneralReg, X86_64FloatReg, ASM, CC>,
+        dst: X86_64GeneralReg,
+        src1: X86_64GeneralReg,
+        src2: X86_64GeneralReg,
+    ) where
+        ASM: Assembler<X86_64GeneralReg, X86_64FloatReg>,
+        CC: CallConv<X86_64GeneralReg, X86_64FloatReg, ASM>,
+    {
+        shift_reg64_reg64_reg64(buf, storage_manager, sar_reg64_reg64, dst, src1, src2)
+    }
+}
+
+fn shift_reg64_reg64_reg64<'a, 'r, ASM, CC>(
+    buf: &mut Vec<'a, u8>,
+    storage_manager: &mut StorageManager<'a, 'r, X86_64GeneralReg, X86_64FloatReg, ASM, CC>,
+    shift_function: fn(buf: &mut Vec<'_, u8>, X86_64GeneralReg),
+    dst: X86_64GeneralReg,
+    src1: X86_64GeneralReg,
+    src2: X86_64GeneralReg,
+) where
+    ASM: Assembler<X86_64GeneralReg, X86_64FloatReg>,
+    CC: CallConv<X86_64GeneralReg, X86_64FloatReg, ASM>,
+{
+    macro_rules! helper {
+        ($buf:expr, $dst:expr, $src1:expr, $src2:expr) => {{
+            mov_reg64_reg64($buf, $dst, $src1);
+            mov_reg64_reg64($buf, X86_64GeneralReg::RCX, $src2);
+
+            shift_function($buf, $dst)
+        }};
+    }
+
+    // if RCX is one of our input registers, we need to move some stuff around
+    if let X86_64GeneralReg::RCX = dst {
```

Comment:
why is this an `if let`? Can't it be `X86_64GeneralReg::RCX == dst` or `matches!(dst, X86_64GeneralReg::RCX)`

Reply:
yes it could be both of those things. This has become my default (often types I work with don't implement `==`, pattern matching always works)